rhapsody/heatmap.py

```diff
@@ -105,6 +106,105 @@ def ranks_heatmap(ranks, microbe_metadata=None, metabolite_metadata=None,
     return hotmap
 
 
+def paired_heatmaps(ranks, microbes_table, metabolites_table, microbe_metadata,
+                    features, top_k_metabolites=50, level=-1,
+                    normalize='log10', color_palette='magma'):
+    '''
+    Creates paired heatmaps of microbe abundances and metabolite abundances.
+
+    Parameters
+    ----------
+    ranks: pd.DataFrame of conditional probabilities.
+        Microbes (rows) X metabolites (columns).
+    microbes_table: biom.Table
+        Microbe feature abundances per sample.
+    metabolites_table: biom.Table
+        Metabolite feature abundances per sample.
+    microbe_metadata: pd.Series
+        Microbe metadata for annotating plots
+    features: list
+        Select microbial feature IDs to display on paired heatmap.
+    top_k_metabolites: int
+        Select top k metabolites associated with the chosen features to
+        display on heatmap.
+    level: int
+        taxonomic level for annotating clustermap.
+        Set to -1 if not parsing semicolon-delimited taxonomies.
+    normalize: str
+        Column normalization strategy to use for heatmaps. Must
+        be "log10", "z_score", or None
+    color_palette: str
+        Color palette for heatmaps.
+    '''
+    # validate microbes
+    missing_microbes = set(features) - set(microbes_table.ids('observation'))
+    if len(missing_microbes) > 0:
+        raise ValueError('features must represent feature IDs in '
+                         'microbes_table. Missing microbe(s): {0}'.format(
+                            missing_microbes))
+
+    # filter select microbes from microbe table and sort by abundance
+    sort_orders = [False] + [True] * (len(features) - 1)
+    select_microbes = microbes_table.copy().filter(features, 'observation')
+    select_microbes = select_microbes.to_dataframe().T
+    select_microbes = select_microbes[features]  # sort cols by features order
+    select_microbes = select_microbes.sort_values(
+        features[::-1], ascending=sort_orders)
+
+    # find top 50 metabolites (highest positive ranks)
+    microb_ranks = ranks.loc[features]
+    top_metabolites = microb_ranks.max()
+    top_metabolites = top_metabolites.sort_values(ascending=False)
+    top_metabolites = top_metabolites[:top_k_metabolites].index
+
+    # grab top 50 metabolites in metabolite table
+    select_metabolites = metabolites_table.copy().filter(
+        top_metabolites, 'observation').to_dataframe().T
+
+    # sort metabolite columns by rank
+    select_metabolites = select_metabolites.reindex(
+        microb_ranks[top_metabolites].sort_values(
+            features[::-1], ascending=sort_orders, axis=1).columns, axis=1)
```

Comment:
@nbokulich is this sorting each column independently?  Wouldn't this destroy the order of the samples?

Reply:
no it is nested sorting, so does not destroy order. E.g.,
```
>>> df = pd.DataFrame({
... 'col1': ['A', 'A', 'A', 'B', 'B', 'B'],
... 'col2': [2, 1, 3, 2, 1, 3],
... 'col3': [0, 1, 9, 4, 2, 3]})
>>> 
>>> df.sort_values(by='col1', ascending=False)
  col1  col2  col3
3    B     2     4
4    B     1     2
5    B     3     3
0    A     2     0
1    A     1     1
2    A     3     9
>>> df.sort_values(by=['col1', 'col2'], ascending=[False, True])
  col1  col2  col3
4    B     1     2
3    B     2     4
5    B     3     3
1    A     1     1
0    A     2     0
2    A     3     9
```